h2o-core/src/main/java/hex/FoldAssignment.java

```diff
@@ -0,0 +1,95 @@
+package hex;
+
+import water.Iced;
+import water.Key;
+import water.exceptions.H2OIllegalArgumentException;
+import water.fvec.Frame;
+import water.fvec.Vec;
+import water.util.ArrayUtils;
+import water.util.VecUtils;
+
+import java.util.Arrays;
+
+public class FoldAssignment extends Iced<FoldAssignment> {
+    final Vec _fold;
+    final int[] _foldMapping;
+    final boolean _internal;
+
+    FoldAssignment(Vec fold, int[] foldMapping, boolean internal) {
+        _fold = fold;
+        _foldMapping = inverseMapping(foldMapping);
```

Comment:
>If internal, then foldMapping is a sequence starting from 0, and inverseMapping will just end up producing the same sequence, but in a non-trivial way.

yes, it is an identity function in that case

> if specified by user, we end up calling inverseMapping(foldMapping(fold)), which seems to mean that we need only one method that produces the "correct" foldMapping directly.

also true, however, I think it is better to separate these 2 into their own methods for testing purposes

I am certainly open to improvements but I also don't see an issue with the current one - it keeps the flow unified for external/internal and even though I am doing stuff I wouldn't have to it doesn't have any impact since the number of folds is small

Reply:
I agree that as we usually keep folds <= 10, the performance impact is negligible.
I'm more concerned about easy it is to reason about the code: because of this separation, the term `foldMapping` represent different things in the `FoldAssignment` class, that's what makes it hard to follow sometimes. If the 2 transformations were applied in the same place, there wouldn't be those different meanings.

Reply:
the naming is confusing, maybe that is the problem - I will try to figure out a better one